### ISSUE TITLE
 How to convert multi-output regression to pmml? It seems that traditional pmml does not support multiple outputs.

### ISSUE THREAD

**Issue by situjunhao** [jpmml/sklearn2pmml#404]

**Comment by vruusmann** [comment]
> How to convert multi-output regression to pmml?

What is the Scikit-Learn estimator type that you're working with? Does the SkLearn2PMML package reject it, or does it generate a PMML document that you simply disapprove (eg. for the reason that it's using a special non-standard XML markup)?

> It seems that traditional pmml does not support multiple outputs.

The support for multiple outputs varies from one PMML model element to another.

For example, the [`RegressionModel`](https://dmg.org/pmml/v4-4-1/Regression.html) element supports any number of target fields, no matter whether you're dealing with regression- or classification-type models. Basically, each target fields is mapped to a separate `RegressionTable` element.

Multi-support is also available with KNN and NaiveBayes models.

TLDR: Please elaborate your use case with a small Python code example and I'll give you more details.

**Comment by situjunhao** [comment]
xgboost.XGBRegressor(
            max_depth=3,
            learning_rate=0.1,
            n_estimators=int(5e2),
            objective='reg:squarederror',  # 多目标
            booster='gbtree',
            gamma=0,
            min_child_weight=1,
            subsample=1,
            colsample_bytree=1,
            reg_alpha=0,
            reg_lambda=1,
            random_state=seed
        )
pipe = PMMLPipeline([("regressor", model)])
pipe.fit(x_train, y_train)
sklearn2pmml(pipe, os.path.join(save_path, f"{model_name}.pmml"), with_repr=True)

The pmml model conversion is successful, but my model has multiple outputs, and I can only get a single output using java calls.

**Comment by vruusmann** [comment]
So this question is really about multi-output XGBoost models (as opposed to native Scikit-Learn models).

This is something that I have on my radar. It requires designing and implementing a "vendor extension" for the PMML standard, that would allow the [`TreeModel`](https://dmg.org/pmml/v4-4-1/TreeModel.html)model element to output multiple scores at once.

Right now, leaf nodes encode the prediction score using the `Node@score` attribute:

```xml
<Node score="0.63317"/>
```

This proposed vendor extension would allow the `Node` element to contain multiple `Score` extension elements instead:
```xml
<Node>
  <Score targetField="y1" value="0.63317"/>
  <Score targetField="y2" value="3.7250"/>
</Node>
```

Perhaps it will be possible to use the existing [`ScoreDistribution`](https://dmg.org/pmml/v4-4-1/TreeModel.html#xsdElement_ScoreDistribution) element for this purpose (ie. no need to define a new `Score` extension element at all). Unfortunately, the `ScoreDistribution` element currently lacks a mechanism for making an association back to a target field (eg. differentiating between `y1` and `y2` fields), so it would be necessary to define a `ScoreDistribution@targetField` extension attribute anyway.

In short, multi-output tree models are not natively supported by PMML schema version 4.4, but they could be enabled pretty easily using a small (JPMML specific-) vendor extension. The same solution would work with Scikit-Learn, XGBoost, and whatever else.

> The pmml model conversion is successful, but my model has multiple outputs, and I can only get a single output using java calls.

I don't see any XGBoost arguments that would indicate that the model is being trained in multi-output mode. Or did XGBoost figure it all by itself, by noticing that your `y_train` dataset has more than one column?

Also, I would have expected the underlying JPMML-XGBoost library to crash in such a scenario (expecting 1-D target, got 2-D or 3-D target).

**Comment by vruusmann** [comment]
Also, there is always the opportunity to "expand" a multi-output XGBoost model into many single-output XGBoost models (so that we could keep using the `Node@score` attribute as-is). Hoewever, the downside of this approach is that we would be duplicating the same tree structure (ie. hierarchies of `Node` elements) over and over again, which drives up memory consumption and doesn't help speed-wise either.

In the future, the SkLearn2PMML package could have a conversion option "style of multi-output tree models". The options would be:

- Use a JPMML-specific vendor extension, and get small and well-performing PMML documents.
- Don't use any extensions, and expand multi-output trees into single-output trees. Will be fully backwards-compatible with legacy systems, but will be much costlier to use.

Also, in principle, it wouldn't be technically difficult to convert between these two representations afterwards (eg. convert one style of PMML documents into the other style, or vice versa).
